Drop rfc6570 uri templates, replace with opcode encoded template.

garretrieger opened the pull request:
Since url templates is primarily a compression mechanism internal to the IFT tables there is no need for templates to be human readable. By switching to an opcode based approach we significantly simplify the template expansion process leading to simpler client implementations.

See: https://github.com/w3c/IFT/issues/259#issuecomment-2794840010 for context and more details.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/IFT/pull/281.html" title="Last updated on Jun 11, 2025, 6:42 PM UTC (03af9d4)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/IFT/281/6e79bc4...03af9d4.html" title="Last updated on Jun 11, 2025, 6:42 PM UTC (03af9d4)">Diff</a>

skef commented on Overview.bs:
```diff
@@ -2016,145 +2021,222 @@ to give the smallest encodings for most unicode code point sets typically encoun
+  <tr><th>Template Bytes</th><th>Input ID</th><th>ID Type</th><th>Expansion</th></tr>
   <tr>
-    <td>//foo.bar/{id}</td>
+    <td>[10, /, /, f, o, o, ., b, a, r, /, 128]</td>
```
Alternatively, we could do something else with the integers, maybe italics. I'm just not entirely comfortable with the "well, there's two characters here so you know what we mean" convention.

garretrieger replied:
Decided to use C style syntax w/ single quotes. A little more cluttered but removes ambiguity and as an added bonus the example inputs can now be directly copied and pasted into code if an implementer wants to use these as test cases.

svgeesus commented on Overview.bs:
```diff
@@ -2016,145 +2021,222 @@ to give the smallest encodings for most unicode code point sets typically encoun
     <td>
-      The second last character of the string in the <code>id</code> variable.
-      If the <code>id</code> variable has less than 2 characters then, the value is the character _ (U+005F).
+      The second last character of the string in the <var>id32</var> variable.
```
Yes, W3C officially uses US English.

Not a blocker, but best to fix.

garretrieger replied:
Fixed.

svgeesus commented on Overview.bs:
```diff
@@ -2016,145 +2021,222 @@ to give the smallest encodings for most unicode code point sets typically encoun
     <td>123</td>
     <td>Integer</td>
     <td>//foo.bar/FC</td>
   </tr>
```
It would. Otherwise we risk some implementation only handling relative urls.

skef replied:
I was actually thinking in the other direction. Maybe we should have at least one of:

* A truly absolute URL
* A protocol-relative URL
* A host-relative URL
* A directory relative URL

Or some subset of these larger than the current examples (which I believe are all protocol-relative).

garretrieger commented on Overview.bs:
```diff
@@ -1087,24 +1087,23 @@ the encoded bytes at the start of every iteration to pick up any changes made in
   <tr>
     <td>uint16</td>
-    <td>uriTemplateLength</td>
+    <td>urlTemplateLength</td>
```
Yeah, I had left URI in this section to be consistent with the RFC we were referencing but now that is no longer used I've switched to only URL throughout the document.

garretrieger commented:
Made some update for review comments. Also added some additional error conditions:
- Op code 0 (which would insert 0 literals) is treated as invalid.
- Error if not enough bytes remain for a literal insert op code.

Also added some examples that include errors.

skef reviewed (approved):
This all looks good to me. I think it would be good to have a directory (as opposed to host)-relatve URL example but I won't hold up this PR for it.

skef commented on Overview.bs:
```diff
@@ -2016,145 +2021,222 @@ to give the smallest encodings for most unicode code point sets typically encoun
+
+     *  Interpret the 7 least significant bits of <var>op code</var> as an unsigned integer, <var>number of literal bytes</var>.
+   
+     *  Read <var>number of literal bytes</var> from <var>url template bytes</var> into <var>literal bytes</var>.
```
Yes, I think the new language is sufficiently clear
